packages/loopring_v3/contracts/impl/agents/FastWithdrawalAgent.sol

```diff
@@ -0,0 +1,204 @@
+/*
+
+  Copyright 2017 Loopring Project Ltd (Loopring Foundation).
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+pragma solidity ^0.5.11;
+pragma experimental ABIEncoderV2;
+
+import "../../iface/IExchangeV3.sol";
+
+import "../../lib/Claimable.sol";
+import "../../lib/ERC20SafeTransfer.sol";
+import "../../lib/MathUint.sol";
+import "../../lib/ReentrancyGuard.sol";
+import "../../lib/SignatureUtil.sol";
+import "../../lib/EIP712.sol";
+
+
+/// @title Fast withdrawal agent implementation.
+///
+///        A mediator/liquidity provider (whoever provides the funds immediately) can call
+///        `executeFastWithdrawals` to provide users immediately with funds onchain.
+///        This allows the security of the funds to be handled by any EOA or smart contract.
+///
+///        Users that want to make use of this functionality have to
+///        authorize this contract as their agent.
+///
+///        TODO: generalize this (here or another agent) with any function call
+///        a user wants to do to be able to simulate completely onchain smart wallet with
+///        funds stored in the Loopring protocol.
+///
+/// @author Brecht Devos - <brecht@loopring.org>
+contract FastWithdrawalAgent is ReentrancyGuard
+{
+    using AddressUtil       for address;
+    using AddressUtil       for address payable;
+    using ERC20SafeTransfer for address;
+    using MathUint          for uint;
+    using SignatureUtil     for bytes32;
+
+    struct FastWithdrawal
+    {
+        address exchange;
+        address from;                   // The owner of the account
+        address to;                     // The address that will receive the tokens withdrawn
+        address token;
+        uint    amount;
+        address feeToken;
+        uint    fee;
+        bool    onchainFeePayment;
+        uint    validUntil;
+        bytes   signature;
+    }
+
+    // Replay protection
+    mapping (bytes32 => bool) txHash;
+
+    // EIP712
+    bytes32 constant public FASTWITHDRAWAL_TYPEHASH = keccak256(
+        "FastWithdrawal(address exchange,address from,address to,address token,uint256 amount,address feeToken,uint256 fee,bool onchainFeePayment,uint256 validUntil)"
+    );
+    bytes32 public DOMAIN_SEPARATOR;
+
+    constructor()
+        public
+    {
+        //DOMAIN_SEPARATOR = EIP712.hash(EIP712.Domain("FastWithdrawalAgent", "1.0", address(this)));
+    }
+
+    // This method is expected to be called by a mediator to facilitate fast withdrawals.
```

Comment:
**It will be great if the mediator doesn't have to stay online to call this method. **

One way to achieve this is to ask each mediator to set up allowance for this Agent, then let the operator call this method by specifying an operator.

If I'm the owner of tokens, I can offer to use my funds, but I don't have to be a business to have my owner server to facilitate the tx.

```
executeFastWithdrawals(...) // only operator
registerMediator(feePercentage); //called by anyone who can provide funds
removeMediator() //called by anyone who want to be removed
```

Reply:
I would do this logic in the mediator contract as explained above so additional security can be built into the smart contract (like spending limits, multi-sig before the call can be made, etc...).

Because it's risky without an agreement with the operator it doesn't seem like people should be registering their contract here without an agreement with the operator, so they may as well share their address offchain.